arelle/PackageManager.py

```diff
@@ -304,46 +304,46 @@ def init(cntlr, loadPackagesConfig=True):
         packagesConfigChanged = False # don't save until something is added to pluginConfig
     pluginMethodsForClasses = {} # dict by class of list of ordered callable function objects
     _cntlr = cntlr
-    
+
 def reset():  # force reloading modules and plugin infos
     packagesConfig.clear()  # dict of loaded module pluginInfo objects by module names
     packagesMappings.clear() # dict by class of list of ordered callable function objects
-    
+
 def orderedPackagesConfig():
     return OrderedDict(
-        (('packages', [OrderedDict(sorted(_packageInfo.items(), 
+        (('packages', [OrderedDict(sorted(_packageInfo.items(),
                                           key=lambda k: {'name': '01',
                                                          'status': '02',
                                                          'version': '03',
                                                          'fileDate': '04',
                                                          'license': '05',
                                                          'URL': '06',
                                                          'description': '07',
-                                                         "publisher": '08', 
+                                                         "publisher": '08',
                                                          "publisherURL": '09',
-                                                         "publisherCountry": '10', 
+                                                         "publisherCountry": '10',
                                                          "publicationDate": '11',
-                                                         "supersededTaxonomyPackages": '12', 
+                                                         "supersededTaxonomyPackages": '12',
                                                          "versioningReports": '13',
                                                          'remappings': '14',
                                                          }.get(k[0],k[0])))
                        for _packageInfo in packagesConfig['packages']]),
          ('remappings',OrderedDict(sorted(packagesConfig['remappings'].items())))))
-    
+
 def save(cntlr):
     global packagesConfigChanged
-    if packagesConfigChanged and cntlr.hasFileSystem:
+    if packagesConfigChanged and cntlr.hasFileSystem and packagesJsonFile is not None:
```

Comment:
This seems wrong. The package config has changed so I'd expect this to create the packages json file if it doesn't already exist, not silently skip the save operation.

Reply:
Hm, what about this instead? We're just reusing the definition of `packagesJsonFile` from the init function.

```
def save(cntlr):
    global packagesConfigChanged
    if packagesConfigChanged and cntlr.hasFileSystem:
        if packagesJsonFile is None:
            packagesJsonFile = packagesJsonFile = cntlr.userAppDir + os.sep + "taxonomyPackages.json"

        with io.open(packagesJsonFile, 'wt', encoding='utf-8') as f:
            jsonStr = _STR_UNICODE(json.dumps(orderedPackagesConfig(), ensure_ascii=False, indent=2)) # might not be unicode in 2.7
            f.write(jsonStr)
        packagesConfigChanged = False
```

Reply:
I'd move the json file line into a function that's also called from init so as not to duplicate the file name definition.

Thinking about this further, you ran into this from the CLI not the GUI, correct? The CLI shouldn't persist packages, so there must be a bug elsewhere that incorrectly caused the save method to be called when it shouldn't have. Do you have a larger stacktrace?

Reply:
Yes, this could potentially be an implementation error on my side. I've implemented Arelle in a local library in this way and the stacktrace is raised on the `model_xbrl.close()` line. Didn't save the rest of it unfortunately but the lines before are just from my local code.

```
try:
    model_manager: ModelManager = ModelManager.initialize(cntlr)
    model_xbrl: ModelXbrl = model_manager.load(
        url_filing, taxonomyPackages=url_taxonomy
    )

    model_roles = extract_model_roles(model_xbrl=model_xbrl)

    model_xbrl.close()
except Exception as exc:
    cntlr.addToLog(
        f"Error {entry.name} due to {exc}",
        level=logging.WARNING,
    )
```

Reply:
Can you try putting the `PackageManager.save` call [in the `Cntlr`](https://github.com/Arelle/Arelle/blob/da93c2046fdfd64aa8c8747b70aa1533cc5d97ad/arelle/Cntlr.py#L392) behind a has gui check?
```Python
if self.hasGui:
    PackageManager.save(self)
```

Reply:
That did the trick - I'm not getting an error after doing that. Perhaps it's cleaner for the commit history to close this PR and open a new one with this change?

Reply:
Rebasing and force-pushing also seems fine and would keep the conversations connected.  (Of course, if a new PR is opened that links to that one, the tracking is still there.)

Reply:
To be honest, I wouldn't know how to properly do such a rebase so I went for the safe option and created a new PR.

Reply:
few options, easiest thing is probably to pull latest master and then blow away your commits with `git reset upstream/master`, `upstream` being the name of your fork which will undo all your commits and then you can just recommit and `git push -f`. Other option is to `git rebase upstream/master` and then `git log` find the commit hash of the commit before your first commit and then `git rebase -i <COMMIT HASH>` and follow the instructions. All that to say that your other branch looks good :)

Reply:
@sagesmith-wf thanks, I've saved your comment here for future reference!